File: solutions/209-minimum-size-subarray-sum/solution.py
from typing import List


class Solution:
    def min_sub_array_len(self, target: int, nums: List[int]) -> int:
        left = 0
        total = 0
        ans = float("inf")

        for right in range(len(nums)):
            total += nums[right]
            while total >= target:
                ans = min(ans, right - left + 1)
                total -= nums[left]
                left += 1

        return 0 if ans == float("inf") else ans


if __name__ == "__main__":
    tests = [
        (7, [2, 3, 1, 2, 4, 3], 2),
        (4, [1, 4, 4], 1),
        (11, [1, 1, 1, 1, 1, 1, 1, 1], 0),
    ]
    sol = Solution()
    for target, nums, expected in tests:
        result = sol.min_sub_array_len(target, nums)
        status = "PASS" if result == expected else "FAIL"
        print(f"{status} | target={target}, nums={nums} => {result} (expected {expected})")

File: solutions/209-minimum-size-subarray-sum/solution.go
package main

import "fmt"

func minSubArrayLen(target int, nums []int) int {
	left := 0
	total := 0
	ans := len(nums) + 1

	for right := 0; right < len(nums); right++ {
		total += nums[right]
		for total >= target {
			if right-left+1 < ans {
				ans = right - left + 1
			}
			total -= nums[left]
			left++
		}
	}

	if ans == len(nums)+1 {
		return 0
	}
	return ans
}

func main() {
	tests := []struct {
		target   int
		nums       []int
		expected int
	}{
		{7, []int{2, 3, 1, 2, 4, 3}, 2},
		{4, []int{1, 4, 4}, 1},
		{11, []int{1, 1, 1, 1, 1, 1, 1, 1}, 0},
	}

	for _, tc := range tests {
		result := minSubArrayLen(tc.target, tc.nums)
		status := "PASS"
		if result != tc.expected {
			status = "FAIL"
		}
		fmt.Printf("%s | target=%d, nums=%v => %d (expected %d)\n", status, tc.target, tc.nums, result, tc.expected)
	}
}
